fix(arithmetic): Support emails in count_if

wmak opened the pull request:
- This is a fix for count_if functions, since the value parameter can be any text value

wmak committed:
fix(arithmetic): Support emails in count_if

- This is just a quick workaround to allow emails in count_if properly,
  there still exists the bigger problem that count_if allows a tag_value
  which is a much wider set of possible characters.

EvanPurkhiser commented on static/app/components/searchSyntax/grammar.pegjs:
```diff
@@ -197,12 +197,12 @@ text_filter
 // Filter keys
 key
-  = value:[a-zA-Z0-9_.-]+ {
+  = value:[a-zA-Z0-9_.@-]+ {
```
🤔 What about `+` in emails. Actually there's a lot of stuff that could be in an email 😬

wmak replied:
I'll add `+` too, but there are a lot of characters here we gotta support to make this technically correct, just going for a quickfix here since count_if will increase visibility on what's lacking on this regex. Will definitely re-visit later 👍

EvanPurkhiser replied:
I feel like.. I would almost rather not add them, since then we'll have to come and clean this stuff up.

Or maybe we should be explicit that this is a temporary thing?

wmak replied:
Hmmm that's a good point, let me see if i can change our function_args to a wildcard instead easily, and if not i'll add some comments explaining that the `@` is temporary

EvanPurkhiser replied:
https://github.com/getsentry/sentry/blob/a7b82be057fefc54aa9f5b0caa525f4f38f8c418/static/app/components/searchSyntax/grammar.pegjs#L219-L223 I would make this use `search_value` instead of `key`

wmak replied:
added new terms in the search grammar to handle quoted values in parameters in d26583b
`quoted_aggregate_param` is based on `quoted_value` since a single regex would consume the trailing `"`, wasn't able to re-use search_value here cause we don't want to replace escaped quotes with plain quotes in a function

wmak committed:
ref: Allow arguments with quotes in the search grammar

- This adds new terms in the search grammar to handle quoted values in
  parameters, the existing use of `key` wasn't sufficient since it
  wasn't covering the actual possible values correctly
- As part of that I also found and fixed a bug with the parse_arguments
  function which wasn't handling quotes properly in some cases

getsantry[bot] committed:
style(lint): Auto commit lint changes

wmak committed:
ref: Updating arithmetic with new terms as well

wmak committed:
reverting some changes unrelated to PR

wmak committed:
leftover changes from first iteration

EvanPurkhiser commented on static/app/components/searchSyntax/grammar.pegjs:
```diff
@@ -217,11 +217,24 @@ aggregate_key
+
+raw_aggregate_param
+  = param:[^()\t\n, \"]+ {
+    return tc.tokenKeySimple(param.join(''), false);
```
In the frontend parser, we should probably just give these real token types instead of just masking them as `keySimple`

I would introduce a new TokenConverter method named `tokenKeyAggregateParam` which would look pretty similar to this one https://github.com/getsentry/sentry/blob/758ff2708ab354b070d79175706428579c860ae5/static/app/components/searchSyntax/parser.tsx#L387-L392 except that it would produce a token type of `Token.KeyAggregateParam`

these two types here would change to use the return from `tokenKeyAggregateParam` https://github.com/getsentry/sentry/blob/758ff2708ab354b070d79175706428579c860ae5/static/app/components/searchSyntax/parser.tsx#L419-L420

EvanPurkhiser commented on static/app/components/searchSyntax/grammar.pegjs:
```diff
@@ -217,11 +217,24 @@ aggregate_key
+
+raw_aggregate_param
+  = param:[^()\t\n, \"]+ {
+    return tc.tokenKeySimple(param.join(''), false);
```
Don't have prettier here, but this actually needs two more spaces in front of it. Sorry for the  nit

wmak replied:
nah, nit away, appreciate it 🙏

though 🤔 is there a way to get prettier in here so I don't have to think about styling? 😁

EvanPurkhiser replied:
There is a prettier plugin.. but it’s kinda.. questionable

EvanPurkhiser commented on src/sentry/api/event_search.py:
```diff
@@ -857,6 +860,18 @@ def visit_aggregate_key(self, node, children):
         return process_list(children[0], children[1])
 
+    def visit_aggregate_param(self, node, children):
+        key = children[0]
```
Let's not name it `key`, maybe `param` or just `return children[0]` 🤷

wmak committed:
ref: Addressing PR comments

- Replacing KeySimple for aggregate params with KeyAggregateParam
- linting grammar

wmak committed:
Merge branch 'master' into wmak/fix/emails-in-count-if

wmak committed:
fix: tests

Zylphrex commented on src/sentry/api/event_search.py:
```diff
@@ -119,22 +119,25 @@
-boolean_value    = ~r"(true|1|false|0)"i &end_value
-text_in_list     = open_bracket text_in_value (spaces comma spaces text_in_value)* closed_bracket &end_value
-numeric_in_list  = open_bracket numeric_value (spaces comma spaces numeric_value)* closed_bracket &end_value
+key                    = ~r"[a-zA-Z0-9_.@-]+"
+quoted_key             = '"' ~r"[a-zA-Z0-9_.@:-]+" '"'
```
Why do `key` and `quoted_key` need `@` in the pattern? Does it not suffice to allow it in `aggregate_param`?

wmak replied:
hm, might be leftover as I was working on this, should be safe to remove :+1:

wmak commented on src/sentry/discover/arithmetic.py:
```diff
@@ -114,13 +114,15 @@ def flatten(remaining):
-field_value          = ~r"[a-zA-Z_\.]+"
+function_name          = ~r"[a-zA-Z_0-9]+"
+numeric_value          = ~r"[+-]?[0-9]+\.?[0-9]*"
+field_value            = ~r"[a-zA-Z_\.]+"
```
ah good reminder let me update that as part of these changes

wmak committed:
ref: Addressing PR comments

wmak committed:
Merge branch 'master' into wmak/fix/emails-in-count-if
